core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala

```diff
@@ -145,83 +189,70 @@ class GroupCoordinator(val brokerId: Int,
                           protocols: List[(String, Array[Byte])],
                           responseCallback: JoinCallback) {
     group.inLock {
-      if (!group.is(Empty) && (!group.protocolType.contains(protocolType) || !group.supportsProtocols(protocols.map(_._1).toSet))) {
-        // if the new member does not support the group protocol, reject it
-        responseCallback(joinError(memberId, Errors.INCONSISTENT_GROUP_PROTOCOL))
-      } else if (group.is(Empty) && (protocols.isEmpty || protocolType.isEmpty)) {
-        //reject if first member with empty group protocol or protocolType is empty
+      if (group.currentState == Dead) {
+        // if the group is marked as dead, it means some other thread has just removed the group
+        // from the coordinator metadata; this is likely that the group has migrated to some other
+        // coordinator OR the group is in a transient unstable phase. Let the member retry
+        // joining without the specified member id.
+        responseCallback(joinError(memberId, Errors.UNKNOWN_MEMBER_ID))
+      } else if (group.protocolNotMatch(protocolType, protocols)) {
         responseCallback(joinError(memberId, Errors.INCONSISTENT_GROUP_PROTOCOL))
-      } else if (memberId != JoinGroupRequest.UNKNOWN_MEMBER_ID && !group.has(memberId)) {
+      } else if (group.isPendingMember(memberId)) {
+        // A rejoining pending member will be accepted.
+        val rejoiningMember = new MemberMetadata(memberId, group.groupId, clientId, clientHost, rebalanceTimeoutMs,
+          sessionTimeoutMs, protocolType, protocols)
+
+        addMemberAndRebalance(rejoiningMember, group, responseCallback)
+        group.removePendingMember(memberId)
+      } else if (!group.has(memberId)) {
         // if the member trying to register with a un-recognized id, send the response to let
-        // it reset its member id and retry
+        // it reset its member id and retry.
         responseCallback(joinError(memberId, Errors.UNKNOWN_MEMBER_ID))
       } else {
         group.currentState match {
-          case Dead =>
```

Comment:
It's a little annoying that we had to make this match non-exhaustive. This will probably generate a warning. Maybe we can add default unexpected case. For example:

```
case unexpected =>
  warn("Attempt to rebalance existing member of group $groupId in unexpected state $unexpected")
  responseCallback(joinError(memberId, Errors.UNKNOWN_MEMBER_ID))
```
I think this could also be used for the `Empty` case (see my comment below).